document-generation/smartdocuments/src/main/kotlin/com/ritense/smartdocuments/service/CamundaSmartDocumentGenerator.kt

```diff
@@ -0,0 +1,92 @@
+ * You may obtain a copy of the License at
+ *
+ * https://joinup.ec.europa.eu/collection/eupl/eupl-text-eupl-12
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" basis,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.ritense.smartdocuments.service
+
+import com.fasterxml.jackson.core.JsonPointer
+import com.ritense.document.domain.Document
+import com.ritense.document.service.DocumentService
+import com.ritense.processdocument.domain.impl.CamundaProcessInstanceId
+import com.ritense.processdocument.service.ProcessDocumentAssociationService
+import org.camunda.bpm.engine.delegate.DelegateExecution
+import org.camunda.bpm.model.bpmn.instance.camunda.CamundaProperties
+import org.springframework.http.MediaType
+
+class CamundaSmartDocumentGenerator(
+    private val smartDocumentGenerator: SmartDocumentGenerator,
+    private val processDocumentAssociationService: ProcessDocumentAssociationService,
+    private val documentService: DocumentService,
+) {
+
+    fun generate(execution: DelegateExecution, templateId: String, mediaTypeValue: String) {
+        val document = getDocument(execution)
+        val templateData = getTemplateData(execution, document)
+        val mediaType = MediaType.valueOf(mediaTypeValue)
+        smartDocumentGenerator.generateAndStoreDocument(document.id(), templateId, templateData, mediaType)
+    }
+
+    private fun getDocument(delegateExecution: DelegateExecution): Document {
+        val processInstanceId = CamundaProcessInstanceId(delegateExecution.processInstanceId)
+        val processDocumentInstance = processDocumentAssociationService.findProcessDocumentInstance(processInstanceId)
+        return if (processDocumentInstance.isPresent) {
+            val jsonSchemaDocumentId = processDocumentInstance.get().processDocumentInstanceId().documentId()
+            documentService.findBy(jsonSchemaDocumentId).orElseThrow()
+        } else {
+            // In case a process has no token wait state ProcessDocumentInstance is not yet created,
+            // therefore out business-key is our last chance which is populated with the documentId also.
+            documentService.get(delegateExecution.businessKey)
+        }
+    }
+
+    private fun getTemplateData(execution: DelegateExecution, document: Document): MutableMap<String, Any> {
+        val camundaPropertiesMap = mutableMapOf<String, Any>()
+        execution
+            .bpmnModelElementInstance
+            .extensionElements
+            .elementsQuery
+            .filterByType(CamundaProperties::class.java)
+            .singleResult()
+            .camundaProperties
+            .associateTo(camundaPropertiesMap) {
+                it.camundaName to getPlaceholderValue(it.camundaValue, execution, document)
```

Comment:
This could result in the same NullPointerException that was part of the red flag stories. It is possible to set a property without setting either the name or the value. If either or both of them are not filled a NPE is thrown.

I created a new story for this problem but maybe you can already fix it at least for this case